mesonbuild/rewriter.py

```diff
@@ -468,21 +500,21 @@ def process_kwargs(self, cmd):
                 arg_node = node.args
         if not node:
             mlog.error('Unable to find the function node')
-        assert(isinstance(node, mparser.FunctionNode))
-        assert(isinstance(arg_node, mparser.ArgumentNode))
+        assert(isinstance(node, FunctionNode))
+        assert(isinstance(arg_node, ArgumentNode))
 
         # Print kwargs info
         if cmd['operation'] == 'info':
             info_data = {}
             for key, val in sorted(arg_node.kwargs.items()):
                 info_data[key] = None
-                if isinstance(val, mparser.ElementaryNode):
+                if isinstance(val, ElementaryNode):
                     info_data[key] = val.value
-                elif isinstance(val, mparser.ArrayNode):
+                elif isinstance(val, ArrayNode):
                     data_list = []
                     for i in val.args.arguments:
                         element = None
-                        if isinstance(i, mparser.ElementaryNode):
+                        if isinstance(i, ElementaryNode):
```

Comment:
I think this should have some indication if it skips data. So the caller knows not to use a "set" operation based on what is returned here.

Reply:
The way it is currently implemented, this should never skip any sources, because the source node resolution is already done [here](https://github.com/mesonbuild/meson/blob/master/mesonbuild/ast/introspection.py#L150-L177). This way the rewriter can simply all unknown nodes.

Reply:
I can get it to produce an empty sources list here when actually executing this with meson sees more sources. Currently just using `+` with two variables that contain arrays of sources does this.
This interface will never be able to reveal all sources in all cases, but i think it should be able to detect common cases where it found all sources. Basically it would set a bool whenever it skips a node while chasing sources. Or maybe we want something richer here anyway and this could be done later.

Reply:
This is a bug in `introspection.py` because `+` nodes are not (yet) resolved there. Traveling the source dependency tree is implemented there and _not_ in the rewriter. The rewriter only gets a list of nodes that (should) already contain all string nodes.

Reply:
#5016 fixes the `+` case.

Reply:
Adding a boolean flag to determine completenes is a TODO for after 0.50.0. See IRC.